cns/deviceplugin/server.go

```diff
@@ -0,0 +1,162 @@
+package deviceplugin
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+const devicePrefix = "NIC-"
+
+type deviceCounter interface {
+	getDeviceCount() int
+}
+
+type Server struct {
+	address             string
+	logger              *zap.Logger
+	deviceCounter       deviceCounter
+	shutdownCh          <-chan struct{}
+	deviceCheckInterval time.Duration
+}
+
+func NewServer(logger *zap.Logger, address string, deviceCounter deviceCounter, deviceCheckInterval time.Duration) *Server {
+	return &Server{
+		address:             address,
+		logger:              logger,
+		deviceCounter:       deviceCounter,
+		shutdownCh:          make(chan struct{}),
```

Comment:
don't make and assign a chan here since we will reassign the shutdownCh with the Run ctx

Reply:
hm this may create an ordering dependency that `Run` **must** be called before `ListAndWatch` so that the channel is not nil 🤔
we might need an additional `started` flag that prevents other methods on the server from being called before it is `Ready`

Reply:
Run will be called starting device plugin gRPC server
Once Ready, device plugin will register with kubelet.
Only once registration is complete with Kubelet, Kubelet will call ListAndWatch and Allocate() etc.
So, I think its fine

Reply:
removed the initial assignment.